mindmeld/components/nlp.py

```diff
@@ -1396,6 +1399,38 @@ def process_query(self, query, allowed_nlp_classes=None, dynamic_resource=None,
             query, dynamic_resource=dynamic_resource, verbose=verbose
         )
 
+        # This code block implements allowed entities described here:
+        # https://github.com/cisco/mindmeld/pull/280
+        if allowed_nlp_classes and all(entity == () for entity in entities):
+            dynamic_gazetteer = dynamic_resource.get(GAZETTEER_RSC, {}) if dynamic_resource else {}
+            entities = []
+            for n_best_query in query:
+                n_best_entities = []
+                for entity in allowed_nlp_classes:
+                    # check if entity is in the gazetteers
+                    if n_best_query.text in {
+                        **self.resource_loader.get_gazetteer(entity)['pop_dict'],
+                        **dynamic_gazetteer.get(entity, {})
+                    }:
+                        span = Span(0, len(n_best_query.text) - 1)
+                        entity = Entity(
+                            text=n_best_query.text,
+                            entity_type=entity
+                        )
+                        query_entity = QueryEntity.from_query(
+                            query=n_best_query, span=span, entity=entity
+                        )
+                        n_best_entities.append(query_entity)
+
+                    # check if entity is a system entity
+                    if entity.startswith(SYSTEM_ENTITY_PREFIX):
+                        for sys_entity in n_best_query.system_entity_candidates:
+                            if sys_entity.entity.type == entity:
+                                n_best_entities.append(sys_entity)
```

Comment:
Is the solution for system entities with no context to just return the entity of that type that duckling is most confident about?

For example, is `allowed_nlp_classes` has `sys_amount-of-money`
```
Query: "$20 5"

System entity candidates:
sys_amount-of-money '$20' 0-2
sys_amount-of-money '$20 5' 0-4
sys_amount-of-money '$20 5' 0-4
```
In this case we would go with `$20`?

Reply:
So currently, without any rule-based help, our entity classifier is identifying the following. This is the currently what will be display with our fix since our fix only gets activate when there are no entities detected by the classifier, but here it detects it:
```
{'text': '$20 5',
 'domain': 'accounts_creditcards',
 'intent': 'transfer_money',
 'entities': [{'text': '5',
   'type': 'sys_amount-of-money',
   'role': None,
   'value': [{'value': 5, 'unit': 'unknown'}],
   'span': {'start': 4, 'end': 4}}]}
```

However, with our change, we detect this:
```
{'text': '$20 5',
 'domain': 'accounts_creditcards',
 'intent': 'transfer_money',
 'entities': [{'text': '$20 5',
   'type': 'sys_amount-of-money',
   'role': None,
   'value': [{'value': 20.05, 'unit': '$'}],
   'span': {'start': 0, 'end': 4}}]}
```